docs/messaging/messaging-spans.md

```diff
@@ -195,21 +197,64 @@ Examples:
-(as opposed to e.g., a queue on which the producer happens to listen): then it should use `SERVER`.
-
 ### Operation names
 
 The following operations related to messages are defined for these semantic conventions:
 
 | Operation name | Description |
 | -------------- | ----------- |
-| `publish`      | A message is sent to a destination by a message producer/client.       |
-| `receive`      | A message is received from a destination by a message consumer/server. |
-| `process`      | A message that was previously received from a destination is processed by a message consumer/server. |
+| `publish`      | One or more messages are provided for publishing to an intermediary. |
+| `create`       | A message is created. |
+| `receive`      | One or more messages are requested by a consumer. |
+| `deliver`      | One or more messages are passed to a consumer. |
+
+### Span kind
+
+[Span kinds](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#spankind)
+SHOULD be set according to the following table, based on the operation a span describes.
+
+| Operation name | Span kind|
+|----------------|-------------|
+| `publish`      | `PRODUCER`, if no `create` spans are present. |
+| `create`       | `PRODUCER`, |
+| `receive`      | `CONSUMER` |
+| `deliver`      | `CONSUMER` |
+
+### Trace structure
+
+#### Producer spans
+
+"Publish" spans SHOULD be created for operations of providing messages for
+sending or publishing to an intermediary. A single "Publish" span can account
+for a single message, or for multiple messages (in the case of providing
+messages in batches). "Create" spans MAY be created. A single "Create" span
+SHOULD account only for a single message. "Create" spans SHOULD either be
+children or links of the related "Publish" span.
+
+If a "Create" span exists for a message, its context SHOULD be injected into
+the message. If no "Create" span exists, the context of the related "Publish"
+span SHOULD be injected into the message.
+
+#### Consumer spans
+
+"Deliver" spans SHOULD be created for operations of passing messages to the
+application when those operations are not initiated by the application code
+(push-based scenarios).
+
+"Receive" spans SHOULD be created for operations of passing messages to the
+application when those operations are initiated by the application code
+(pull-based scenarios).
+
+"Deliver" or "Receive" spans MUST NOT be created for messages that are
+pre-fetched or cached by messaging libraries or SDKs until they are forwarded
+to the caller.
+
+A single "Deliver" or "Receive" span can account for a single message, for
+multiple messages (in case messages are passed for processing as batches), or
+for no message at all (if it is signalled that no messages were received). For
+each message it accounts for, the "Deliver" or "Receive" span SHOULD link to
+the message's creation context. In addition, if it is possible the creation
```

Comment:
Although rather informal/verbose in the OTEP itself, I'd like to get some justification/guideline on this one. From the OTEP:

> For single-message scenarios, and if the "Deliver" or "Receive" spans would be root spans of a new trace, the creation context may also be used as a parent on those operations in addition to being added as a link. Keeping single-messages operations in the same trace can greatly improve the user experience.

Otherwise this may be used as de facto backdoor to not use `Links` (better be safe than sorry).

Reply:
@carlosalberto 

> Otherwise this may be used as de facto backdoor to not use Links (better be safe than sorry).

We always require links to be present. Parent-child relationship can be created _in addition_, where it's possible.

I'd prefer to not elaborate too much on the parent-child relationship option (as I fear that might lead people astray), but I'm happy to add clarifications if you think it's necessary.